Implement configurable handling of excess tokens in datasets

mhenrichsen opened the pull request:
- Added `excess_token_handling` option to the configuration, allowing users to choose between "drop" and "truncate" for handling tokens exceeding the maximum sequence length.
- Introduced `truncate_or_drop_long_seq` function to manage both single and batched samples based on the selected handling method.
- Updated relevant dataset processing functions to utilize the new handling option, ensuring backward compatibility with existing "drop" behavior.
- Enhanced logging to reflect truncation actions in dataset processing.

This change improves flexibility in managing sequence lengths during training and evaluation.


# Description
Adds `excess_token_handling`. Either drops or truncates longer sequences.

## Motivation and Context


## How has this been tested?
Ran the following config:
```
base_model: NousResearch/Llama-3.2-1B
# Automatically upload checkpoint and final model to HF
# hub_model_id: username/custom_model_name

datasets:
  - path: teknium/GPT4-LLM-Cleaned
    type: alpaca
dataset_prepared_path: last_run_prepared
val_set_size: 0.1
output_dir: ./outputs/lora-out

adapter: lora
lora_model_dir:

sequence_len: 256
sample_packing: true
eval_sample_packing: true
pad_to_sequence_len: true
excess_token_handling: truncate

lora_r: 16
lora_alpha: 32
lora_dropout: 0.05
lora_target_modules:
  - gate_proj
  - down_proj
  - up_proj
  - q_proj
  - v_proj
  - k_proj
  - o_proj

wandb_project:
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 2
micro_batch_size: 2
num_epochs: 1
optimizer: adamw_8bit
lr_scheduler: cosine
learning_rate: 0.0002

bf16: auto
tf32: false

gradient_checkpointing: true
resume_from_checkpoint:
logging_steps: 1
flash_attention: true

loss_watchdog_threshold: 5.0
loss_watchdog_patience: 3

warmup_steps: 10
evals_per_epoch: 4
saves_per_epoch: 1
weight_decay: 0.0
special_tokens:
  pad_token: "<|end_of_text|>"
  ```
  
  With `excess_token_handling: truncate` it gives:
  ```
  [2025-05-12 12:32:25,508] [DEBUG] [axolotl.calculate_total_num_steps:624] [PID:5957] [RANK:0] total_num_steps: 805
[2025-05-12 12:32:25,556] [DEBUG] [axolotl.calculate_total_num_steps:513] [PID:5957] [RANK:0] total_num_tokens: 7_449_409
[2025-05-12 12:32:25,856] [DEBUG] [axolotl.calculate_total_num_steps:531] [PID:5957] [RANK:0] `total_supervised_tokens: 4_856_047`
```

And with `drop`:
```
[RANK:0] total_num_steps: 593
[2025-05-12 12:13:11,503] [DEBUG] [axolotl.calculate_total_num_steps:513] [PID:691] [RANK:0] total_num_tokens: 4_549_075
[2025-05-12 12:13:11,747] [DEBUG] [axolotl.calculate_total_num_steps:531] [PID:691] [RANK:0] `total_supervised_tokens: 2_405_977`
```



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a configuration option to control how over-length sequences are handled: choose to drop them or truncate to fit.
  * Truncation support applied across pretraining and RL preference datasets, preserving samples by trimming responses when needed.
* **Tests**
  * Added extensive test coverage for sequence overflow handling in both general and RL data flows, including edge cases and batching.
* **Style**
  * Minor log message formatting improvements for optimizer-related output.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mhenrichsen committed:
Implement configurable handling of excess tokens in datasets

- Added `excess_token_handling` option to the configuration, allowing users to choose between "drop" and "truncate" for handling tokens exceeding the maximum sequence length.
- Introduced `truncate_or_drop_long_seq` function to manage both single and batched samples based on the selected handling method.
- Updated relevant dataset processing functions to utilize the new handling option, ensuring backward compatibility with existing "drop" behavior.
- Enhanced logging to reflect truncation actions in dataset processing.

This change improves flexibility in managing sequence lengths during training and evaluation.

mhenrichsen committed:
lint

mhenrichsen committed:
Refactor truncation logic in `drop_long_rl_seq` function

- Simplified the truncation process for chosen and rejected responses to ensure they fit within the specified sequence length while preserving the prompt.
- Improved readability by restructuring the code and removing redundant checks.
- Ensured that the function returns the sample correctly after processing, maintaining compatibility with existing handling options.

mhenrichsen committed:
fix linting issues

winglian commented on docs/config.qmd:
```diff
@@ -332,6 +332,8 @@ dataset_shard_idx:
 # as most models have a token/context limit of 2048
 sequence_len: 2048
+# How to handle tokens exceeding max sequence length - "drop" (default, removes sample) or "truncate" (cuts off excess tokens)
+excess_token_handling: drop
```
maybe something like `sequence_len_excess` or `sequence_len_overflow` would be obvious

winglian commented on src/axolotl/utils/data/pretraining.py:
```diff
@@ -259,6 +259,8 @@ def encode_packed_pretraining(
         # workaround by using the position id logic for now in trainer
         drop_attention_mask=multipack_attn,
+        # pass through handling mode from config via ds_wrapper function
+        handling=getattr(ds_wrapper, "cfg", {}).get("excess_token_handling", "drop"),
```
maybe somewhere in constants we have a DEFAULT_{whatever we name_it}_VALUE = "drop" and can reference it throughout

mhenrichsen replied:
Where would you want this placed?

mhenrichsen replied:
@winglian we just need this one then we are good to go I think

NanoCode012 replied:
@mhenrichsen , let's put it in the adjacent `utils.py`?

winglian commented on src/axolotl/utils/data/rl.py:
```diff
@@ -96,11 +98,51 @@ def drop_long_rl_seq(
+            ) <= sequence_len
+
+        # truncate
+        else:
```
since the original default is drop, my preference is `if truncate; do; else` so the fallback is always the drop functionality.

winglian commented on src/axolotl/utils/trainer.py:
```diff
@@ -370,15 +478,31 @@ def drop_no_trainable_tokens(sample):
+            desc="Dropping Long Sequences",
+            load_from_cache_file=False,
+        )
+    else:
```
Is this needed if the new function handles both?

mhenrichsen replied:
You're right, the separate filter logic isn't needed anymore. I've updated `process_pretraining_datasets_for_packing` to use the same `truncate_or_drop_long_seq` function for both modes. It now returns a boolean when `handling="drop"` (for use with `.filter()`) and the modified sample when `handling="truncate"` (for use with `.map()`), consolidating the logic as you suggested.

winglian reviewed (commented):
Would be good to add a unit test that runs a dataset through and checks the lengths of the outputs and counts (drop should have fewer rows whereas truncate should have the same count as the input dataset)

mhenrichsen committed:
Fixes comments from winglian

mhenrichsen committed:
Linting fix

mhenrichsen committed:
docstring for tests

mhenrichsen committed:
lint

mhenrichsen committed:
increased test coverage

mhenrichsen committed:
fix merge conflicts

mhenrichsen committed:
Merge branch 'main' of github.com:axolotl-ai-cloud/axolotl into 775-option-to-drop-vs-truncate-on-rows-longer-than-context-length

mhenrichsen committed:
Refactor sequence length overflow handling in pretraining module

- Introduced DEFAULT_SEQUENCE_LEN_OVERFLOW_HANDLING constant in utils.py.
- Updated encode_packed_pretraining function to use this constant instead of a hardcoded value.

mhenrichsen committed:
Merge branch 'main' into 775-option-to-drop-vs-truncate-on-rows-longer-than-context-length

mhenrichsen committed:
RL datasets: warn and drop unsalvageable over-length prompts post-truncate; add post-truncate filter; support alias config key 'excess_token_handling'

mhenrichsen committed:
Resolve merge conflicts: unify pretraining utils imports, add alias handling; fix rl.py per new RL dataset API; resolve config schema conflict and add sequence_len_overflow_handling field

mhenrichsen committed:
Clean up conflict markers; finalize RL data split implementation; fix config schema conflicts; add truncation+post-filter behavior and alias handling

mhenrichsen committed:
Refactor AxolotlInputConfig: clean up sequence_len and sequence_len_overflow_handling fields, ensuring consistent descriptions and removing conflict markers.

mhenrichsen committed:
remove unused files

mhenrichsen committed:
pre-commit: fix rl.py imports/types; add legacy drop_long_rl_seq wrapper; resolve config schema; run formatting

mhenrichsen committed:
Remove transscribe.py file and clean up optimizer.py and rl.py for improved formatting and consistency.

mhenrichsen committed:
Remove deprecated configuration files: deleted config.qmd and finetune copy.yml to streamline project structure and eliminate unused resources.
